File: server/src/main/java/com/nutricare/service/impl/AssessmentService.java
package com.nutricare.service.impl;

import com.nutricare.domain.entity.*;
import com.nutricare.domain.enums.PredictionStatus;
import com.nutricare.domain.enums.StuntStatus;
import com.nutricare.dto.request.assessment.AssessmentRequest;
import com.nutricare.dto.response.prediction.PredictionResponse;
import com.nutricare.dto.response.PageResponse;
import com.nutricare.exception.ForbiddenException;
import com.nutricare.exception.ResourceNotFoundException;
import com.nutricare.repository.*;
import com.nutricare.util.CuidGenerator;
import com.nutricare.util.ZScoreCalculator;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.Pageable;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.Period;
import java.util.List;

/**
 * AssessmentService — Phase 3
 * BE-302: POST /api/assessments (simpan + trigger prediksi async)
 * BE-304: GET  /api/assessments/{id}
 * BE-305: GET  /api/assessments/child/{childId}
 * BE-306: Retry job untuk prediksi PENDING
 */
@Service
@RequiredArgsConstructor
@Slf4j
public class AssessmentService {

    private final AssessmentRepository assessmentRepository;
    private final PredictionRepository predictionRepository;
    private final ChildRepository childRepository;
    private final BlockchainAnchorRepository blockchainAnchorRepository;
    private final ZScoreCalculator zScoreCalculator;
    private final PredictionService predictionService;

    // ── BE-302: Buat assessment baru ──────────────────────────────────────────

    @Transactional
    public PredictionResponse createAssessment(AssessmentRequest request, String userId) {
        // Verifikasi anak milik user
        Child child = childRepository.findByIdAndUserId(request.getChildId(), userId)
            .orElseThrow(() -> new ForbiddenException("Anak tidak ditemukan atau bukan milik Anda"));

        // Hitung usia anak dalam bulan
        int ageMonths = Period.between(child.getBirthDate(), LocalDate.now()).getYears() * 12
                      + Period.between(child.getBirthDate(), LocalDate.now()).getMonths();

        // Hitung Z-Score di server (bukan di AI — sesuai aturan klinis CONTEXT.md)
        BigDecimal zscoreHa = zScoreCalculator.calculateHeightForAge(
            request.getHeight().doubleValue(), ageMonths, child.getGender());
        BigDecimal zscoreWa = zScoreCalculator.calculateWeightForAge(
            request.getWeight().doubleValue(), ageMonths, child.getGender());
        BigDecimal zscoreWh = zScoreCalculator.calculateWeightForHeight(
            request.getWeight().doubleValue(), request.getHeight().doubleValue(), child.getGender());

        // Tentukan status stunting
        StuntStatus status = zScoreCalculator.determineStuntStatus(zscoreHa);
        short riskLevel = zScoreCalculator.determineRiskLevel(status);

        // Simpan assessment (immutable)
        Assessment assessment = Assessment.builder()
            .id(CuidGenerator.generate())
            .child(child)
            .weight(request.getWeight())
            .height(request.getHeight())
            .headCircumference(request.getHeadCircumference())
            .bfExclusive(request.getBfExclusive())
            .mpasiAge(request.getMpasiAge())
            .mealFreq(request.getMealFreq())
            .illnessHistory(request.getIllnessHistory())
            .build();
        assessmentRepository.save(assessment);

        // Buat prediction dengan status PENDING
        Prediction prediction = Prediction.builder()
            .id(CuidGenerator.generate())
            .assessment(assessment)
            .status(status)
            .predictionStatus(PredictionStatus.PENDING)
            .zscoreHa(zscoreHa)
            .zscoreWa(zscoreWa)
            .zscoreWh(zscoreWh)
            .riskLevel(riskLevel)
            .build();
        predictionRepository.save(prediction);

        // Trigger Gemini secara async (tidak memblokir response)
        predictionService.generatePredictionAsync(prediction.getId(), assessment, child, ageMonths);

        return mapToResponse(prediction, child);
    }

    // ── BE-304: Get assessment by ID ──────────────────────────────────────────

    public PredictionResponse getAssessment(String assessmentId, String userId) {
        Assessment assessment = assessmentRepository.findById(assessmentId)
            .orElseThrow(() -> new ResourceNotFoundException("Assessment tidak ditemukan"));

        // Ownership check untuk PARENT
        if (!assessment.getChild().getUser().getId().equals(userId)) {
            throw new ForbiddenException("Anda tidak memiliki akses ke data ini");
        }

        Prediction prediction = predictionRepository.findByAssessmentId(assessmentId)
            .orElseThrow(() -> new ResourceNotFoundException("Prediksi tidak ditemukan"));

        return mapToResponse(prediction, assessment.getChild());
    }

    // ── BE-305: Get assessments by child ─────────────────────────────────────

    public PageResponse<PredictionResponse> getAssessmentsByChild(
            String childId, String userId, Pageable pageable) {

        childRepository.findByIdAndUserId(childId, userId)
            .orElseThrow(() -> new ForbiddenException("Anak tidak ditemukan atau bukan milik Anda"));

        org.springframework.data.domain.Page<com.nutricare.domain.entity.Assessment> pageResult = assessmentRepository.findByChildIdOrderByCreatedAtDesc(childId, pageable);
        
        java.util.List<PredictionResponse> responses = pageResult.stream()
            .map(a -> {
                Prediction p = predictionRepository.findByAssessmentId(a.getId()).orElse(null);
                return p != null ? mapToResponse(p, a.getChild()) : null;
            })
            .toList();

        return PageResponse.<PredictionResponse>builder()
            .data(responses)
            .page(pageResult.getNumber())
            .size(pageResult.getSize())
            .totalElements(pageResult.getTotalElements())
            .totalPages(pageResult.getTotalPages())
            .build();
    }

    // ── BE-306: Retry job untuk PENDING ──────────────────────────────────────

    @Scheduled(fixedDelay = 300000) // tiap 5 menit
    public void retryPendingPredictions() {
        List<Prediction> pending = predictionRepository.findByPredictionStatus(PredictionStatus.PENDING);
        log.info("Retry job: {} prediksi PENDING ditemukan", pending.size());
        for (Prediction prediction : pending) {
            Assessment assessment = prediction.getAssessment();
            Child child = assessment.getChild();
            int ageMonths = Period.between(child.getBirthDate(), LocalDate.now()).getYears() * 12
                          + Period.between(child.getBirthDate(), LocalDate.now()).getMonths();
            predictionService.generatePredictionAsync(prediction.getId(), assessment, child, ageMonths);
        }
    }

    // ── Helper ────────────────────────────────────────────────────────────────

    private PredictionResponse mapToResponse(Prediction prediction, Child child) {
        blockchainAnchorRepository.findByAssessmentId(prediction.getAssessment().getId())
            .ifPresent(anchor -> {});

        var anchor = blockchainAnchorRepository
            .findByAssessmentId(prediction.getAssessment().getId())
            .map(a -> PredictionResponse.BlockchainInfo.builder()
                .anchorStatus(a.getAnchorStatus().name())
                .txHash(a.getTxHash())
                .polygonscanUrl(a.getTxHash() != null
                    ? "https://amoy.polygonscan.com/tx/" + a.getTxHash()
                    : null)
                .isVerified(a.getBlockNumber() != null)
                .build())
            .orElse(null);

        return PredictionResponse.builder()
            .id(prediction.getId())
            .assessmentId(prediction.getAssessment().getId())
            .childId(child.getId())
            .childName(child.getName())
            .status(prediction.getStatus())
            .predictionStatus(prediction.getPredictionStatus())
            .riskLevel(prediction.getRiskLevel())
            .statusLabel(PredictionResponse.getLabelForStatus(prediction.getStatus()))
            .statusColor(PredictionResponse.getColorForStatus(prediction.getStatus()))
            .zscoreHa(prediction.getZscoreHa())
            .zscoreWa(prediction.getZscoreWa())
            .zscoreWh(prediction.getZscoreWh())
            .summary(prediction.getSummary())
            .nextAssessmentDate(prediction.getNextAssessmentDate())
            .disclaimer(PredictionResponse.DISCLAIMER)
            .blockchain(anchor)
            .createdAt(prediction.getCreatedAt())
            .build();
    }
}

File: mobile/src/features/assessment/services/assessment-service.ts
import { supabase } from '@/utils/supabase';
import type { PageResponse } from '@/types/api-types';
import type {
  AssessmentRequestDTO,
  AssessmentResponseDTO,
  AssessmentPredictionDTO,
  BlockchainAnchorDTO,
} from '@/features/assessment/types/assessment-types';
import type { StuntStatus } from '@/features/children/types/child-types';

// ── Supabase row types ────────────────────────────────────────────────────

type SupabaseAssessmentRow = {
  id: string;
  child_id: string;
  weight: number | null;
  height: number | null;
  head_circumference: number | null;
  bf_exclusive: boolean | null;
  mpasi_age: number | null;
  meal_freq: number | null;
  illness_history: string | null;
  created_at: string;
};

type SupabasePredictionRow = {
  id: string;
  assessment_id: string;
  stunt_status: StuntStatus;
  prediction_status: 'PENDING' | 'COMPLETED' | 'FAILED';
  zscore_wa: number | null;
  zscore_ha: number | null;
  zscore_wh: number | null;
  risk_level: number | null;
  summary: string | null;
  recommendations: string[] | null;
  next_assessment_date: string | null;
  ai_limited: boolean | null;
  created_at: string;
};

// ── Transform helpers ────────────────────────────────────────────────────

function toPrediction(row: SupabasePredictionRow | null): AssessmentPredictionDTO {
  if (!row) {
    return {
      id: '',
      status: 'NORMAL' as StuntStatus,
      predictionStatus: 'PENDING',
      zscoreWa: 0,
      zscoreHa: 0,
      zscoreWh: 0,
      riskLevel: 0,
      summary: '',
      recommendations: [],
      nextAssessmentDate: '',
      disclaimer: '',
      aiLimited: false,
    };
  }
  return {
    id: row.id,
    status: row.stunt_status,
    predictionStatus: row.prediction_status,
    zscoreWa: row.zscore_wa ?? 0,
    zscoreHa: row.zscore_ha ?? 0,
    zscoreWh: row.zscore_wh ?? 0,
    riskLevel: row.risk_level ?? 0,
    summary: row.summary ?? '',
    recommendations: row.recommendations ?? [],
    nextAssessmentDate: row.next_assessment_date ?? '',
    disclaimer: '',
    aiLimited: row.ai_limited ?? false,
  };
}

function toAssessment(
  row: SupabaseAssessmentRow & { prediction?: SupabasePredictionRow | null },
  child?: { id: string; name: string }
): AssessmentResponseDTO {
  return {
    id: row.id,
    child: child ?? { id: row.child_id, name: '' },
    weight: row.weight ?? undefined,
    height: row.height ?? undefined,
    headCircumference: row.head_circumference ?? undefined,
    bfExclusive: row.bf_exclusive ?? undefined,
    mpasiAge: row.mpasi_age ?? undefined,
    mealFreq: row.meal_freq ?? undefined,
    illnessHistory: row.illness_history ?? undefined,
    createdAt: row.created_at,
    prediction: toPrediction(row.prediction ?? null),
  };
}

// ── Service ──────────────────────────────────────────────────────────────

/**
 * Supabase returns ONE-TO-MANY relations as arrays even when cardinality is 1.
 * prediction:predictions(*) → [{...}] not {...}. Normalize here.
 */
function firstOrNull<T>(val: T | T[] | null | undefined): T | null {
  if (val == null) return null;
  return Array.isArray(val) ? (val[0] ?? null) : val;
}

export const assessmentService = {
  createAssessment: async (data: AssessmentRequestDTO): Promise<AssessmentResponseDTO> => {
    const { data: inserted, error } = await supabase
      .from('assessments')
      .insert({
        child_id: data.childId,
        weight: data.weight,
        height: data.height,
        head_circumference: data.headCircumference || null,
        bf_exclusive: data.bfExclusive,
        mpasi_age: data.mpasiAge || null,
        meal_freq: data.mealFreq,
        illness_history: data.illnessHistory || null,
      })
      .select()
      .single();
    if (error) throw error;
    return toAssessment(inserted as unknown as SupabaseAssessmentRow, { id: data.childId, name: '' });
  },

  getAssessment: async (assessmentId: string): Promise<AssessmentResponseDTO> => {
    const { data, error } = await supabase
      .from('assessments')
      .select(`
        *,
        prediction:predictions(*)
      `)
      .eq('id', assessmentId)
      .single();
    if (error) throw error;
    // Supabase returns prediction as array (one-to-many) — normalize to single row
    const row = data as unknown as SupabaseAssessmentRow & { prediction: SupabasePredictionRow[] | SupabasePredictionRow | null };
    return toAssessment({ ...row, prediction: firstOrNull(row.prediction) });
  },

  getPrediction: async (assessmentId: string): Promise<AssessmentPredictionDTO> => {
    const { data, error } = await supabase
      .from('predictions')
      .select('*')
      .eq('assessment_id', assessmentId)
      .single();
    if (error) {
      return {
        id: '', status: 'NORMAL', predictionStatus: 'PENDING',
        zscoreWa: 0, zscoreHa: 0, zscoreWh: 0, riskLevel: 0,
        summary: '', recommendations: [], nextAssessmentDate: '', disclaimer: '',
      };
    }
    return toPrediction(data as unknown as SupabasePredictionRow);
  },

  getChildAssessments: async (childId: string, page = 0, size = 10): Promise<PageResponse<AssessmentResponseDTO>> => {
    const from = page * size;
    const to = from + size - 1;
    const { data, error, count } = await supabase
      .from('assessments')
      .select(`
        *,
        prediction:predictions(*)
      `, { count: 'exact' })
      .eq('child_id', childId)
      .order('created_at', { ascending: false })
      .range(from, to);
    if (error) throw error;
    const rows = (data ?? []) as unknown as (SupabaseAssessmentRow & { prediction: SupabasePredictionRow[] | SupabasePredictionRow | null })[];
    return {
      data: rows.map((r) => toAssessment({ ...r, prediction: firstOrNull(r.prediction) })),
      page,
      size,
      totalElements: count ?? 0,
      totalPages: count ? Math.ceil(count / size) : 0,
    };
  },
};
